xmtp_mls/src/groups/mod.rs

```diff
@@ -551,31 +551,53 @@ where
                 )))?;
             let group = Self::new(context, group.id, group.dm_id, group.created_at_ns);
 
+            tracing::warn!("Skipping old welcome {}", welcome.id);
             return Ok(group);
         };
 
-        let mut decrypted_welcome: Option<Result<DecryptedWelcome, GroupError>> = None;
-        let result = provider.transaction(|provider| {
-            let result = DecryptedWelcome::from_encrypted_bytes(
+        let mut decrypt_result: Result<DecryptedWelcome, GroupError> =
+            Err(GroupError::UninitializedResult);
+        let transaction_result = provider.transaction(|provider| {
+            decrypt_result = DecryptedWelcome::from_encrypted_bytes(
                 provider,
                 &welcome.hpke_public_key,
                 &welcome.data,
                 welcome.wrapper_algorithm.into(),
             );
-            decrypted_welcome = Some(result);
             Err(StorageError::IntentionalRollback)
         });
 
         // TODO: Move cursor forward on non-retriable errors, but not on retriable errors
-        let Err(StorageError::IntentionalRollback) = result else {
-            return Err(result?);
+        let Err(StorageError::IntentionalRollback) = transaction_result else {
+            return Err(transaction_result?);
         };
 
-        let DecryptedWelcome { staged_welcome, .. } = decrypted_welcome.expect("Set to some")?;
-
+        let DecryptedWelcome { staged_welcome, .. } = decrypt_result?;
         // Ensure that the list of members in the group's MLS tree matches the list of inboxes specified
         // in the `GroupMembership` extension.
         validate_initial_group_membership(&context, &staged_welcome).await?;
+        let group_id = staged_welcome.public_group().group_id();
+        if provider.db().find_group(group_id.as_slice())?.is_some() {
+            // Fetch the original MLS group, rather than the one from the welcome
+            let (group, _) = MlsGroup::new_cached(context.clone(), group_id.as_slice())?;
```

Comment:
We are sure that in the case that the group came in from DeviceSync this will always at least return an `Ok` result?

Does the DeviceSync MLS group think that it is active? Is it in epoch 0?